### PR TITLE
Object classification by scale (it narrows down the candidates)

### LINKED ISSUE #2123: Scale (体重計) under the storage system
DATASHEET: https://www.aandd.co.jp/adhome/pdf/manual/balance/ekew-i.pdf (

### PR DESCRIPTION
**Pull request by pazeshun**
Close #2123 

## How to use
```
$ roslaunch jsk_arc2017_common object_estimation_by_scale.launch

# Get current weight (the sum of 4 scales)
$ rostopic echo /estimate_object_by_scale/weight_sum
```
~~# Set current state as initial state
$ rostopic pub -1 /estimate_object_by_scale/initialize std_msgs/Empty~~
```
# Set current state as initial state
$ rosservice call /estimate_object_by_scale/initialize

# Get candidates of an object picked from initial state
$ rostopic echo /estimate_object_by_scale/picked_object_candidates

# Get candidates of an object stowed from initial state
$ rostopic echo /estimate_object_by_scale/stowed_object_candidates
```

## Movie
https://drive.google.com/open?id=0BxxBA3J-CunGYzV5cGMzdkN1R28

### CODE REVIEW

File: jsk_arc2017_common/node_scripts/estimate_object_by_scale.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            '~initialize', Empty, self._initialize)
+        self.lock = threading.Lock()
+
+    def subscribe(self):
+        self.scale_subs = []
+        for i, scale_input in enumerate(self.scale_inputs):
+            self.scale_subs.append(
+                rospy.Subscriber(
+                    scale_input, Float32, self._scale_cb, callback_args=i)
+                )
```
Comment (knorth55) [inline jsk_arc2017_common/node_scripts/estimate_object_by_scale.py]:
```
self.scale_subs.append(
    rospy.Subscriber(
        scale_input, Float32, self._scale_cb, callback_args=i))
```

or

```
self.scale_subs.append(
    rospy.Subscriber(
        scale_input, Float32, self._scale_cb, callback_args=i)
)
```

File: jsk_arc2017_common/node_scripts/estimate_object_by_scale.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+        self.weight_sum_pub.publish(Float32(weight_sum))
+        self.picked_pub.publish(picked_object)
+        self.stowed_pub.publish(stowed_object)
+        self.lock.release()
+
+    def _initialize(self, req):
+        self.lock.acquire()
+        self.init_sum = sum(self.scale_values)
+        self.lock.release()
```
Comment (knorth55) [inline jsk_arc2017_common/node_scripts/estimate_object_by_scale.py]: Does this thread locking really work?
I think you need a `Thread` instance to lock.
https://github.com/start-jsk/jsk_apc/blob/master/jsk_arc2017_baxter/node_scripts/state_server.py

Comment (pazeshun) [inline jsk_arc2017_common/node_scripts/estimate_object_by_scale.py]: Is ROS subscriber in Python not implemented with Thread?
I'm not sure...

File: jsk_arc2017_common/node_scripts/estimate_object_by_scale.py (hunk 1 above)
Comment (pazeshun) [inline jsk_arc2017_common/node_scripts/estimate_object_by_scale.py]: Fixed. Thx

File: jsk_arc2017_common/node_scripts/estimate_object_by_scale.py (hunk 2 above)
Comment (wkentaro) [inline jsk_arc2017_common/node_scripts/estimate_object_by_scale.py]: I think it looks good.

Comment (knorth55) [inline jsk_arc2017_common/node_scripts/estimate_object_by_scale.py]: ok

**Comment by pazeshun** [comment]
Updated movie!

File: jsk_arc2017_common/launch/object_estimation_by_scale.launch [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    <rosparam>port: /dev/ttyUSB0</rosparam>
+  </node>
+  <node name="scale1" pkg="jsk_arc2017_common" type="ekew_i_driver.py">
+    <rosparam>port: /dev/ttyUSB1</rosparam>
+  </node>
+  <node name="scale2" pkg="jsk_arc2017_common" type="ekew_i_driver.py">
+    <rosparam>port: /dev/ttyUSB2</rosparam>
+  </node>
+  <node name="scale3" pkg="jsk_arc2017_common" type="ekew_i_driver.py">
+    <rosparam>port: /dev/ttyUSB3</rosparam>
```
Comment (pazeshun) [inline jsk_arc2017_common/launch/object_estimation_by_scale.launch]: I think those scales should be pluged into new desktop PC.
If those scales are pluged into sheeta, we have to make udev rules to make symlink like Arduino

**Comment by pazeshun** [comment]
Please merge this

**Comment by pazeshun** [comment]
Oh, thx!

**Comment by wkentaro** [comment]
Nice work! :tada:

**Comment by k-okada** [comment]
ask @furushchev for how to use bayesian process to estimate the likeliness of the object, when you write the paper.

**Pull request merged**